File: src/main/scala/net/machinemuse/utils/render/GuiIcons.scala
package net.machinemuse.utils.render

import net.minecraft.util.IIcon
import net.machinemuse.powersuits.common.Config
import org.lwjgl.opengl.GL11._
import net.machinemuse.numina.geometry.Colour
import net.machinemuse.numina.render.{MuseTextureUtils, RenderState, MuseIconUtils}

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 4:01 AM, 30/04/13
 */
object GuiIcons {

  trait GuiIcon {
    val filepath: String
    val size: Double

    def apply(x: Double, y: Double, c: Colour = Colour.WHITE, xmin: Double = Integer.MIN_VALUE, ymin: Double = Integer.MIN_VALUE, xmax: Double = Integer.MAX_VALUE, ymax: Double = Integer.MAX_VALUE) {
      MuseTextureUtils.pushTexture(filepath)
      glPushMatrix()
      RenderState.blendingOn()
      val s = size / 16.0
      glScaled(s, s, s)
      MuseIconUtils.drawIconPartialOccluded(x / s, y / s, GuiIconDrawer, c, xmin / s, ymin / s, xmax / s, ymax / s)
      RenderState.blendingOff()
      glPopMatrix()
      MuseTextureUtils.popTexture()
    }
  }

  object Checkmark extends GuiIcon {
    val size = 16.0
    val filepath = Config.TEXTURE_PREFIX + "gui/checkmark.png"
  }

  object TransparentArmor extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/transparentarmor.png"
  }

  object NormalArmor extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/normalarmor.png"
  }

  object GlowArmor extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/glowarmor.png"
  }

  object SelectedArmorOverlay extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/armordisplayselect.png"
  }

  object ArmourColourPatch extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/colourclicker.png"
  }

  object MinusSign extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/minussign.png"
  }

  object PlusSign extends GuiIcon {
    val size = 8.0
    val filepath = Config.TEXTURE_PREFIX + "gui/plussign.png"
  }

  object GuiIconDrawer extends IIcon {
    def getOriginX: Int = 0

    def getOriginY: Int = 0

    def getMinU: Float = 0

    def getMaxU: Float = 1

    def getInterpolatedU(d0: Double): Float = d0.toFloat

    def getMinV: Float = 0

    def getMaxV: Float = 1

    def getInterpolatedV(d0: Double): Float = d0.toFloat

    def getIconName: String = "GuiIcon"

    def getIconWidth: Int = 8

    def getIconHeight: Int = 8

  }

}

File: src/main/java/net/machinemuse/utils/render/GuiIcons.java
package net.machinemuse.utils.render;

import net.machinemuse.numina.geometry.Colour;
import net.machinemuse.numina.render.MuseIconUtils;
import net.machinemuse.numina.render.MuseTextureUtils;
import net.machinemuse.numina.render.RenderState;
import net.machinemuse.powersuits.common.Config;
import net.minecraft.util.IIcon;
import org.lwjgl.opengl.GL11;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 4:01 AM, 30/04/13
 *
 * Ported to Java by lehjr on 10/19/16.
 */
public class GuiIcons {
        public static class GuiIcon {

        final double size;
        final String filepath;

        final double x;
        final double y;
        final Colour c;
        final double xmin;
        final double ymin;
        final double xmax;
        final double ymax;

        public GuiIcon (double size, String filepath, double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            this.size = size;
            this.filepath = filepath;
            this.x = x;
            this.y = y;
            this.c = (c != null) ? c : Colour.WHITE;
            this.xmin = (xmin != null) ? xmin : Integer.MIN_VALUE;
            this.ymin = (ymin != null) ? ymin : Integer.MIN_VALUE;
            this.xmax = (xmax != null) ? xmax : Integer.MAX_VALUE;
            this.ymax = (ymax != null) ? ymax : Integer.MAX_VALUE;

            MuseTextureUtils.pushTexture(filepath);
            GL11.glPushMatrix();
            RenderState.blendingOn();
            double s = size / 16.0;
            GL11.glScaled(s, s, s);

            MuseIconUtils.drawIconPartialOccluded(x / s, y / s, new GuiIconDrawer(), this.c, this.xmin / s, this.ymin / s, this.xmax / s, this.ymax / s);
            RenderState.blendingOff();
            GL11.glPopMatrix();
            MuseTextureUtils.popTexture();
        }
    }

    public static class Checkmark extends GuiIcon {
        public Checkmark(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(16.0, Config.TEXTURE_PREFIX + "gui/checkmark.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class TransparentArmor extends GuiIcon {
        public TransparentArmor(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/transparentarmor.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class NormalArmor extends GuiIcon {
        public NormalArmor(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/normalarmor.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class GlowArmor extends GuiIcon {
        public GlowArmor(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/glowarmor.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class SelectedArmorOverlay extends GuiIcon {
        public SelectedArmorOverlay(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/armordisplayselect.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class ArmourColourPatch extends GuiIcon {
        public ArmourColourPatch(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/colourclicker.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class MinusSign extends GuiIcon {
        public MinusSign(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/minussign.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class PlusSign extends GuiIcon {
        public PlusSign(double x, double y, Colour c, Double xmin, Double ymin, Double xmax, Double ymax) {
            super(8.0, Config.TEXTURE_PREFIX + "gui/plussign.png", x, y, c, xmin, ymin, xmax, ymax);
        }
    }

    public static class GuiIconDrawer implements IIcon {
        @Override
        public int getIconWidth() {
            return 8;
        }

        @Override
        public int getIconHeight() {
            return 8;
        }

        @Override
        public float getMinU() {
            return 0;
        }

        @Override
        public float getMaxU() {
            return 1;
        }

        @Override
        public float getInterpolatedU(double d0) {
            return (float) d0;
        }

        @Override
        public float getMinV() {
            return 0;
        }

        @Override
        public float getMaxV() {
            return 1;
        }

        @Override
        public float getInterpolatedV(double d0) {
            return (float) d0;
        }

        @Override
        public String getIconName() {
            return "GuiIcon";
        }

        public int getOriginX() {
            return 0;
        }

        public int getOriginY() {
            return 0;
        }
    }
}